File: src/main/java/org/goodmath/pica/util/Twist.java
package org.goodmath.pica.util;

import java.util.ArrayList;
import java.util.List;

public abstract class Twist implements Twistable {
    public abstract void render(int curIndent, StringBuilder target);

    @Override
    public Twist twist() {
        return this;
    }

    private static class ObjectNode extends Twist {
        private String name;
        private List<Twistable> children;
        public ObjectNode(String name, List<Twistable> children) {
            this.name = name;
            this.children = children;
        }

        @Override
        public void render(int i, StringBuilder sb) {
            indent(i, sb);
            sb.append("obj ").append(name).append(":\n");
            for(Twistable child: children) {
                child.twist().render(i+1, sb);
            }
        }
    }

    private static class AttributeNode extends Twist {
        private String name;
        private String value;
        public AttributeNode(String name, String value) {
            this.name = name;
            this.value = value;
        }

        @Override
        public void render(int curIndent, StringBuilder target) {
            indent(curIndent, target);
            target.append(name).append(" = ").append(value).append("\n");
        }
    }

    private static class ValueNode extends Twist {
        private final String name;
        private final Twistable value;

        public ValueNode(String name, Twistable value) {
            this.name = name;
            this.value = value;
        }

        @Override
        public void render(int curIndent, StringBuilder target) {
            indent(curIndent, target);
            target.append(name).append(":\n");
            value.twist().render(curIndent + 1, target);
        }
    }

    private static class ArrayNode extends Twist {
        private final String name;
        private final List<? extends Twistable> values;
        public ArrayNode(String name, List<? extends Twistable> children) {
            this.name = name;
            this.values = children;
        }

        public ArrayNode(String name, Twistable... children) {
            this(name, List.of(children));
        }

        @Override
        public void render(int curIndent, StringBuilder target) {
            if (!values.isEmpty()) {
                indent(curIndent, target);
                target.append(name).append(" = array:\n");
                for (Twistable t : values) {
                    t.twist().render(curIndent + 1, target);
                }
            }
        }
    }

    public static Twist obj(String name, List<Twistable> children) {
        return new ObjectNode(name, children);
    }

    public static Twist obj(String name, Twistable... children) {
        return new ObjectNode(name, List.of(children));
    }

    public static Twist attr(String name, String val) {
        return new AttributeNode(name, val);
    }

    public static Twist val(String name, Twistable value) {
        return new ValueNode(name, value);
    }

    public static Twist arr(String name, List<? extends Twistable> values) {
        return new ArrayNode(name, values);
    }

    public static Twist arr(String name, Twistable... values) {
        return new ArrayNode(name, List.of(values));
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        render(0, sb);
        return sb.toString();
    }

    protected void indent(int count, StringBuilder sb) {
        sb.append("   ".repeat(count));
    }

}

File: src/main/kotlin/org/goodmath/pica/util/twist.kt
/* Copyright 2022, Mark C. Chu-Carroll
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.goodmath.pica.util

fun StringBuilder.indent(i: Int) {
    append("   ".repeat(i))
}

interface Twistable {
    fun twist(): Twist
}

sealed class Twist: Twistable {
    override fun toString(): String {
        val sb = StringBuilder()
        this.render(sb, 0)
        return sb.toString()
    }



    override fun twist(): Twist = this
    abstract fun render(sb: StringBuilder, indent: Int)

    abstract fun renderCode(i: Int): String
    companion object {
        fun obj(name: String, vararg children: Twistable): Twist {
            return ObjectNode(name, children.toList())
        }

        fun obj(name: String, children: List<Twistable>): Twist {
            return ObjectNode(name, children)
        }

        fun arr(name: String, vararg children: Twistable): Twist {
            return ArrayNode(name, children.toList())
        }

        fun arr(name: String, children: List<Twistable>): Twist {
            return ArrayNode(name, children)
        }

        fun attr(name: String, value: String): Twist {
            return AttributeNode(name, value)
        }

        fun value(name: String, value: Twistable): Twist {
            return ValueNode(name, value)
        }

        fun leaf(value: String): Twist {
            return LeafNode(value)
        }

        fun opt(t: Twistable?): Twist {
            return OptNode(t)
        }
    }
}

class ObjectNode(val name: String, val children: List<Twistable>): Twist() {
    override fun render(sb: StringBuilder, indent: Int) {
        sb.indent(indent)
        sb.append("obj $name:\n")
        for (c in children) {
            c.twist().render(sb, indent + 1)
        }
    }

    override fun renderCode(i: Int): String =
        "  ".repeat(i) +
                "Twist.obj(\"${name}\",\n" +
                children.map { it.twist().renderCode(i+1) }.joinToString("") +
                "  ".repeat(i) + ")\n"
}

class ArrayNode(val name: String, val children: List<Twistable>): Twist() {
    override fun render(sb: StringBuilder, indent: Int) {
        if (children.isNotEmpty()) {
            sb.indent(indent)
            sb.append("$name = array:\n")
            for (c in children) {
                c.twist().render(sb, indent + 1)
            }
        }
    }

    override fun renderCode(i: Int): String {
        return if (children.isNotEmpty()) {
            "  ".repeat(i) +
                    "Twist.arr(\"${name}\",\n" +
                    children.map { it.twist().renderCode(i + 1) }.joinToString("") +
                    "  ".repeat(i) + ")\n"
        } else {
            ""
        }
    }
}

class AttributeNode(val name: String, val value: String): Twist() {
    override fun render(sb: StringBuilder, indent: Int) {
        sb.indent(indent)
        sb.append("$name = $value\n")
    }

    override fun renderCode(i: Int): String {
        return "  ".repeat(i) + "Twist.attr(\"$name\", \"$value\"),\n"
    }
}

class ValueNode(val name: String, val value: Twistable): Twist(){
    override fun render(sb: StringBuilder, indent: Int) {
        sb.indent(indent)
        sb.append("$name:\n")
        value.twist().render(sb, indent+1)
    }

    override fun renderCode(i: Int): String {
        return "  ".repeat(i) + "Twist.value(\"$name\",\n" + value.twist().renderCode(i+1) +
                "  ".repeat(i) +
                ")\n"
    }
}

class LeafNode(val name: String): Twist() {
    override fun render(sb: StringBuilder, indent: Int) {
        sb.indent(indent)
        sb.append(name).append("\n")
    }

    override fun renderCode(i: Int): String {
        return "  ".repeat(i) + "Twist.leaf(\"$name\")\n"
    }
}

class OptNode(val t: Twistable?): Twist() {
    override fun render(sb: StringBuilder, indent: Int) {
        t?.twist()?.render(sb, indent)
    }

    override fun renderCode(i: Int): String {
        return t?.twist()?.renderCode(i) ?: ""
    }

}